### ISSUE TITLE
[Style dropdown] Block style applied to the block before a block widget

### ISSUE THREAD

**Issue by niegowski** [ckeditor/ckeditor5#11585]
## 📝 Provide detailed reproduction steps (if any)

![2022-04-07 11 15 16](https://user-images.githubusercontent.com/1232187/162165417-1e513efd-80a5-44f4-81f7-7e80bcce4499.gif)

### ✔️ Expected result

Only blocks that are visually selected should be affected.

### ❌ Actual result

A block before a block widget gets the style.

## 📃 Other details

* Browser: …
* OS: …
* First affected CKEditor version: …
* Installed CKEditor plugins: …

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by filipsobol** [comment]
This seems to be consistent with the behavior of other block operations.

If we select the last paragraph, the image, and the end of the heading, as in the GIF above, the style, quote, and code block operations **will affect the heading block**.
However, if we select only the last paragraph and image, the style and code block operations are disabled, and the quote operation **will not affect the heading block**.

See this video where I first select only the last paragraph and image, and then also the end of the heading.


https://user-images.githubusercontent.com/4145208/229166084-b65047e9-9280-4106-b875-b5333f3e4517.mp4

**Comment by Witoso** [comment]
Good find @filipsobol! I just want to share another recording that may better show where the problem is, and why it may be confusing for the user:

https://user-images.githubusercontent.com/9881379/229437817-3f904c2d-4eb5-4b3d-a3cc-2529ba5304d3.mov

Notice that as a user I only slightly raise my selection above the paragraph. My expectation is I select part of the paragraph, and the image. But this small selection (user's perspective), selects heading as well. It's even more visible if you have an empty paragraph. Users may have the intuition, "I only select what's on the level of my cursor", and unfortunately we break this pattern here.

Not sure where the problem lies but it's still there.

**Comment by filipsobol** [comment]
Other editors have better visual indications when end of the block is selected.

Word:
https://user-images.githubusercontent.com/4145208/229442680-9c895652-0b1b-473f-95f7-b4219fb972b7.mp4

Google docs:
https://user-images.githubusercontent.com/4145208/229442759-fd66d178-ef6b-4365-85b2-809eddb0f815.mp4

**Comment by Witoso** [comment]
> There should be an exception if this block is empty. Such blocks are still treated as selected:

I'm a bit worried about this one. And not sure if I understand it well.

```
<heading2>[</heading2> // Selection starting at the beginning of the empty block counts
<imageBlock></imageBlock>
<paragraph>c]</paragraph>
```

But this case in style dropdown will be also confusing. It will be more visible (I will see a highlight of the selection) but the selection will still be applied when I move the cursor just above the paragraph. Right?

![image](https://user-images.githubusercontent.com/9881379/229530220-f416a64c-aaba-4dc8-8a1b-c29e47b9bb92.png)

**Comment by filipsobol** [comment]
Yes, currently that's the case. Interestingly, it happens on Chrome, but not in Firefox.


https://user-images.githubusercontent.com/4145208/229535000-03277892-476e-4dc3-9495-aa14791e36da.mp4
